# Data import and descriptive analysis

In [43]:
#Import librearies
import pandas as pd
import numpy as np
import os

## Import the order dataset

In [2]:
df = pd.read_excel(r'D:\PythonPrjkt\NTSB_Aviation_accident_Analysis\02_Data\NTSB_Aviation_Truncated.xlsx', index_col = False)

In [3]:
pd.set_option("display.max_columns", None)

In [4]:
df.head()

ev_id  Aircraft_Key              Aircraft_Mfg_Mdl_srs  \
0  20080107X00026             1                  PIPER-PA 28-180-   
1  20080107X00026             1                  PIPER-PA 28-180-   
2  20080107X00026             2          Barnard/Stancil-Glastar-   
3  20080107X00026             2          Barnard/Stancil-Glastar-   
4  20080109X00036             1  Micco Aircraft Company-MAC-145B-   

  acft_category  total_seats  num_eng date_last_insp  \
0           AIR          4.0      1.0     2007-04-01   
1           AIR          4.0      1.0     2007-04-01   
2           AIR          2.0      1.0     2007-01-01   
3           AIR          2.0      1.0     2007-01-01   
4           AIR          2.0      1.0     2007-09-01   

   Airframe_Hrs_since_last_inspection  Airframe_hrs_total        oper_name  \
0                                75.0              2865.0  Timothy Bennett   
1                                75.0              2865.0  Timothy Bennett   
2                                 NaN               340.0    Kenneth Couey   
3                                 NaN               340.0    Kenneth Couey   
4                                 NaN               516.0              NaN   

                        USA_State   oper_country oper_pax_cargo type_fly  \
0  CALIFORNIA                      United States            NaN     PERS   
1  CALIFORNIA                      United States            NaN     PERS   
2  CALIFORNIA                      United States            NaN     PERS   
3  CALIFORNIA                      United States            NaN     PERS   
4                             NaN            NaN            NaN     PERS   

  ev_type    ev_date Day Time of Event  ev_year  ev_month mid_air latitude  \
0     ACC 2008-01-01  Tu      22:30:00     2008         1       Y  381326N   
1     ACC 2008-01-01  Tu      22:30:00     2008         1       Y  381326N   
2     ACC 2008-01-01  Tu      22:30:00     2008         1       Y  381326N   
3     ACC 2008-01-01  Tu      22:30:00     2008         1       Y  381326N   
4     ACC 2008-01-01  Tu      22:00:00     2008         1     NaN  293022N   

  longitude             Airport Nearest_Airport On or off Airport  apt_dist  \
0  1222659W  Schellville/Sonoma             OQ3              ONAP       0.0   
1  1222659W  Schellville/Sonoma             OQ3              ONAP       0.0   
2  1222659W  Schellville/Sonoma             OQ3              ONAP       0.0   
3  1222659W  Schellville/Sonoma             OQ3              ONAP       0.0   
4  0952836W   HOUSTON-SOUTHWEST             AXH              ONAP       0.0   

  light_cond Sky_condition  Runway_Visibility_Feet  Visibility Miles  \
0       DAYL          CLER                     NaN              10.0   
1       DAYL          CLER                     NaN              10.0   
2       DAYL          CLER                     NaN              10.0   
3       DAYL          CLER                     NaN              10.0   
4       DAYL          CLER                     NaN              10.0   

   gust_kts  Total Fatalities  Total Injuries Investigating_Agency  \
0       0.0                 0             3.0                    N   
1       0.0                 0             3.0                    N   
2       0.0                 0             3.0                    N   
3       0.0                 0             3.0                    N   
4      20.0                 0             1.0                    N   

  Weather_Condition  dec_latitude  dec_longitude  \
0               VMC     38.223888    -122.449722   
1               VMC     38.223888    -122.449722   
2               VMC     38.223888    -122.449722   
3               VMC     38.223888    -122.449722   
4               VMC     29.506111     -95.476669   

                                     Events_Sequence engines Type  \
0                           Landing Midair collision          REC   
1                     Initial climb Midair collision          REC   
2               

## Check shape

In [5]:
df.shape

(43989, 41)

## Check data types

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43989 entries, 0 to 43988
Data columns (total 41 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   ev_id                               43989 non-null  object        
 1   Aircraft_Key                        43989 non-null  int64         
 2   Aircraft_Mfg_Mdl_srs                43989 non-null  object        
 3   acft_category                       43989 non-null  object        
 4   total_seats                         39247 non-null  float64       
 5   num_eng                             41437 non-null  float64       
 6   date_last_insp                      34193 non-null  datetime64[ns]
 7   Airframe_Hrs_since_last_inspection  13672 non-null  float64       
 8   Airframe_hrs_total                  33305 non-null  float64       
 9   oper_name                           22511 non-null  object        
 10  USA_State             

## Check duplicates

In [7]:
dfdups = df[df.duplicated()] # find
#df_prods_clean_no_dups = df_prods_clean.drop_duplicates() #  remove

In [8]:
dfdups.shape

(0, 41)

### No duplicates found

In [9]:
df.describe()

Aircraft_Key   total_seats       num_eng  \
count  43989.000000  39247.000000  41437.000000   
mean       1.014981      6.267001      1.113063   
std        0.121478     83.667507      0.335185   
min        1.000000      1.000000      0.000000   
25%        1.000000      2.000000      1.000000   
50%        1.000000      4.000000      1.000000   
75%        1.000000      4.000000      1.000000   
max        2.000000   8300.000000      4.000000   

       Airframe_Hrs_since_last_inspection  Airframe_hrs_total       ev_year  \
count                        1.367200e+04        33305.000000  43989.000000   
mean                         2.800729e+03         4666.119400   2014.072518   
std                          1.798487e+05         8795.669076      3.872698   
min                         -3.300000e+01            0.000000   2008.000000   
25%                          9.300000e+00         1095.000000   2011.000000   
50%                          2.500000e+01         3222.899902   2014.000000   
75%                          5.900000e+01         5568.669922   2017.000000   
max                          1.214202e+07       725421.000000   2022.000000   

           ev_month      apt_dist  Runway_Visibility_Feet  Visibility Miles  \
count  43989.000000  43948.000000               61.000000      40117.000000   
mean       6.619950     23.037726             2612.836066         11.438884   
std        3.009519    355.725100             1823.564305         94.176679   
min        1.000000     -1.000000               -2.000000          0.000000   
25%        4.000000      0.000000              800.000000         10.000000   
50%        7.000000      0.000000             2000.000000         10.000000   
75%        9.000000      0.000000             4000.000000         10.000000   
max       12.000000   7671.000000             6000.000000       9999.000000   

           gust_kts  Total Fatalities  Total Injuries  dec_latitude  \
count  43948.000000      43989.000000    43740.000000  4.284000e+04   
mean       3.048398          0.465776        4.439186  2.244858e+03   
std       12.110376          3.977647       24.164596  3.501983e+05   
min        0.000000          0.000000        0.000000 -4.857139e+01   
25%        0.000000          0.000000        1.000000  3.352694e+01   
50%        0.000000          0.000000        2.000000  3.825778e+01   
75%        0.000000          0.000000        2.000000  4.242389e+01   
max      999.000000        295.000000      576.000000  6.777778e+07   

       dec_longitude  
count   4.284100e+04  
mean    2.180225e+02  
std     5.966387e+04  
min    -1.230857e+06  
25%    -1.129664e+02  
50%    -9.437139e+01  
75%    -8.219029e+01  
max     1.222222e+07

## All columns look ok apart from the following
###  total_seats  max value cannot be 8300
###  Airframe_Hrs_since_last_inspection	 , one negative value observed, cannot be negative
###  apt_dist , one negative value and max value of 7671 miles from the nearest airport does not look right
###  Runway_Visibility_Feet , one negative value,cannot be negative
###  Visibility Miles, max value 9999 seems unrealistic
### gust_kts ,max value 999 seems  unrealistic 




####  total_seats max value cannot be 8300
This air craft has max 11 seats. Value needs imputing
https://en.wikipedia.org/wiki/Beechcraft_Model_18

In [10]:
df.loc[df['total_seats'] == 8300]

ev_id  Aircraft_Key Aircraft_Mfg_Mdl_srs acft_category  \
5342  20090729X15804             1          BEECH-H-18-           AIR   
5343  20090729X15804             1          BEECH-H-18-           AIR   
5344  20090729X15804             1          BEECH-H-18-           AIR   

      total_seats  num_eng date_last_insp  Airframe_Hrs_since_last_inspection  \
5342       8300.0      2.0     2009-07-10                                 NaN   
5343       8300.0      2.0     2009-07-10                                 NaN   
5344       8300.0      2.0     2009-07-10                                 NaN   

      Airframe_hrs_total oper_name                       USA_State  \
5342              6780.0     Pilot  MISSOURI                         
5343              6780.0     Pilot  MISSOURI                         
5344              6780.0     Pilot  MISSOURI                         

       oper_country oper_pax_cargo type_fly ev_type    ev_date Day  \
5342  United States            NaN     PERS     ACC 2009-07-24  Fr   
5343  United States            NaN     PERS     ACC 2009-07-24  Fr   
5344  United States            NaN     PERS     ACC 2009-07-24  Fr   

     Time of Event  ev_year  ev_month mid_air latitude longitude  \
5342      23:12:00     2009         7     NaN  393520N  0841329W   
5343      23:12:00     2009         7     NaN  393520N  0841329W   
5344      23:12:00     2009         7     NaN  393520N  0841329W   

                             Airport Nearest_Airport On or off Airport  \
5342  Dayton-Wright Brothers Airport             MGY              ONAP   
5343  Dayton-Wright Brothers Airport             MGY              ONAP   
5344  Dayton-Wright Brothers Airport             MGY              ONAP   

      apt_dist light_cond Sky_condition  Runway_Visibility_Feet  \
5342       0.0       DAYL           NaN                     NaN   
5343       0.0       DAYL           NaN                     NaN   
5344       0.0       DAYL           NaN                     NaN   

      Visibility Miles  gust_kts  Total Fatalities  Total Injuries  \
5342               NaN       0.0                 0             4.0   
5343               NaN       0.0                 0             4.0   
5344               NaN       0.0                 0             4.0   

     Investigating_Agency Weather_Condition  dec_latitude  dec_longitude  \
5342                    N               VMC      39.58889     -84.224723   
5343                    N               VMC      39.58889     -84.224723   
5344                    N               VMC      39.58889     -84.224723   

                                        Events_Sequence engines Type  \
5342  Landing-flare/touchdown Landing gear not confi...          REC   
5343     Landing-landing roll Loss of control on ground          REC   
5344  Landing-landing roll Collision with terr/obj (...          REC   

     engine_Manufacturer  
5342     Pratt & Whitney  
5343     Pratt & Whitney  
5344     Pratt & Whitney

In [11]:
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

date_last_insp
oper_name
USA_State
oper_country
oper_pax_cargo
type_fly
Time of Event
mid_air
latitude
longitude
Airport
Nearest_Airport
On or off Airport
light_cond
Sky_condition
Investigating_Agency
Weather_Condition
Events_Sequence
engines Type
engine_Manufacturer


### total_seats max value cannot be 8300. The max seats are 853 on an Airbus A380
I will replace any value greater than 853 by NaN

In [12]:
df['total_seats'].values[df['total_seats'].values > 859] = np.nan

In [13]:
df['total_seats'].describe()

count    39237.000000
mean         5.140097
std         18.362891
min          1.000000
25%          2.000000
50%          4.000000
75%          4.000000
max        495.000000
Name: total_seats, dtype: float64

### Airframe_Hrs_since_last_inspection , one negative value observed, cannot be negative. I will replace it with Nan

In [14]:
df['Airframe_Hrs_since_last_inspection'].values[df['Airframe_Hrs_since_last_inspection'].values < 0] = np.nan

In [15]:
df['Airframe_Hrs_since_last_inspection'].describe()

count    1.367100e+04
mean     2.800937e+03
std      1.798553e+05
min      0.000000e+00
25%      9.300000e+00
50%      2.500000e+01
75%      5.900000e+01
max      1.214202e+07
Name: Airframe_Hrs_since_last_inspection, dtype: float64

In [16]:
df.loc[df['apt_dist'] == 7671]

ev_id  Aircraft_Key Aircraft_Mfg_Mdl_srs acft_category  \
33178  20180125X54359             1      PIPER-PA32R-300           AIR   
33179  20180125X54359             1      PIPER-PA32R-300           AIR   
33180  20180125X54359             1      PIPER-PA32R-300           AIR   
33181  20180125X54359             1      PIPER-PA32R-300           AIR   

       total_seats  num_eng date_last_insp  \
33178          6.0      NaN     2017-10-20   
33179          6.0      NaN     2017-10-20   
33180          6.0      NaN     2017-10-20   
33181          6.0      NaN     2017-10-20   

       Airframe_Hrs_since_last_inspection  Airframe_hrs_total oper_name  \
33178                                50.0         6156.859863       NaN   
33179                                50.0         6156.859863       NaN   
33180                                50.0         6156.859863       NaN   
33181                                50.0         6156.859863       NaN   

                            USA_State   oper_country oper_pax_cargo type_fly  \
33178  FLORIDA                         United States            NaN     PERS   
33179  FLORIDA                         United States            NaN     PERS   
33180  FLORIDA                         United States            NaN     PERS   
33181  FLORIDA                         United States            NaN     PERS   

      ev_type    ev_date Day Time of Event  ev_year  ev_month mid_air  \
33178     ACC 2018-01-25  Th      19:25:00     2018         1     NaN   
33179     ACC 2018-01-25  Th      19:25:00     2018         1     NaN   
33180     ACC 2018-01-25  Th      19:25:00     2018         1     NaN   
33181     ACC 2018-01-25  Th      19:25:00     2018         1     NaN   

      latitude longitude                         Airport Nearest_Airport  \
33178  245725N  0008132W  THE FLORIDA KEYS MARATHON INTL             MTH   
33179  245725N  0008132W  THE FLORIDA KEYS MARATHON INTL             MTH   
33180  245725N  0008132W  THE FLORIDA KEYS MARATHON INTL             MTH   
33181  245725N  0008132W  THE FLORIDA KEYS MARATHON INTL             MTH   

      On or off Airport  apt_dist light_cond Sky_condition  \
33178               NaN    7671.0       DAYL           NaN   
33179               NaN    7671.0       DAYL           NaN   
33180               NaN    7671.0       DAYL           NaN   
33181               NaN    7671.0       DAYL           NaN   

       Runway_Visibility_Feet  Visibility Miles  gust_kts  Total Fatalities  \
33178                     NaN              10.0       0.0                 0   
33179                     NaN              10.0       0.0                 0   
33180                     NaN              10.0       0.0                 0   
33181                     NaN              10.0       0.0                 0   

       Total Injuries Investigating_Agency Weather_Condition  dec_latitude  \
33178             4.0                    N               VMC     24.956943   
33179             4.0                    N               VMC     24.956943   
33180             4.0                    N               VMC     24.956943   
33181             4.0                    N               VMC     24.956943   

       dec_longitude                              Events_Sequence  \
33178     -81.050552            Takeoff Loss of control on ground   
33179     -81.050552       Takeoff Attempted remediation/recovery   
33180     -81.050552                     Takeoff Runway excursion   
33181     -81.050552  Takeoff Collision with terr/obj (non-CFIT)    

      engines Type engine_Manufacturer  
33178          REC            Lycoming  
33179          REC            Lycoming  
33180          REC            Lycoming  
33181          REC            Lycoming

### This crash happened during take off on the same airport therefore this value of 7671 is definitely incorrect. replacing it with Nan

In [17]:
df['apt_dist'].values[df['apt_dist'].values == 7671 ] = np.nan

In [18]:
df['apt_dist'].describe()

count    43944.000000
mean        22.341571
std        348.176808
min         -1.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       7470.000000
Name: apt_dist, dtype: float64

In [19]:
df['apt_dist'].value_counts().loc[lambda x : x > 5000]

0.0    34964
1.0     5084
Name: apt_dist, dtype: int64

In [20]:
df.loc[df['apt_dist'] == 7470]

ev_id  Aircraft_Key             Aircraft_Mfg_Mdl_srs  \
29772  20170212X02618             1  NIELSEN MARTIN M-CHALLENGER II-   
29773  20170212X02618             1  NIELSEN MARTIN M-CHALLENGER II-   
29774  20170212X02618             1  NIELSEN MARTIN M-CHALLENGER II-   

      acft_category  total_seats  num_eng date_last_insp  \
29772           AIR          2.0      1.0     2016-11-06   
29773           AIR          2.0      1.0     2016-11-06   
29774           AIR          2.0      1.0     2016-11-06   

       Airframe_Hrs_since_last_inspection  Airframe_hrs_total oper_name  \
29772                                 NaN           22.299999       NaN   
29773                                 NaN           22.299999       NaN   
29774                                 NaN           22.299999       NaN   

                            USA_State   oper_country oper_pax_cargo type_fly  \
29772  FLORIDA                         United States            NaN     PERS   
29773  FLORIDA                         United States            NaN     PERS   
29774  FLORIDA                         United States            NaN     PERS   

      ev_type    ev_date Day Time of Event  ev_year  ev_month mid_air  \
29772     ACC 2017-02-11  Sa      22:40:00     2017         2     NaN   
29773     ACC 2017-02-11  Sa      22:40:00     2017         2     NaN   
29774     ACC 2017-02-11  Sa      22:40:00     2017         2     NaN   

      latitude longitude      Airport Nearest_Airport On or off Airport  \
29772  265446N  0815931W  PUNTA GORDA             PGD              ONAP   
29773  265446N  0815931W  PUNTA GORDA             PGD              ONAP   
29774  265446N  0815931W  PUNTA GORDA             PGD              ONAP   

       apt_dist light_cond Sky_condition  Runway_Visibility_Feet  \
29772    7470.0       DAYL           UNK                     NaN   
29773    7470.0       DAYL           UNK                     NaN   
29774    7470.0       DAYL           UNK                     NaN   

       Visibility Miles  gust_kts  Total Fatalities  Total Injuries  \
29772               NaN       0.0                 0             1.0   
29773               NaN       0.0                 0             1.0   
29774               NaN       0.0                 0             1.0   

      Investigating_Agency Weather_Condition  dec_latitude  dec_longitude  \
29772                    N               VMC     26.912776     -81.991943   
29773                    N               VMC     26.912776     -81.991943   
29774                    N               VMC     26.912776     -81.991943   

                                   Events_Sequence engines Type  \
29772            Landing Loss of control in flight          REC   
29773       Landing Attempted remediation/recovery          REC   
29774  Landing Collision with terr/obj (non-CFIT)           REC   

      engine_Manufacturer  
29772               Rotax  
29773               Rotax  
29774               Rotax

In [21]:
df['apt_dist'].values[df['apt_dist'].values < 0 ] = np.nan

In [22]:
df['apt_dist'].describe()

count    43927.000000
mean        22.350604
std        348.243873
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       7470.000000
Name: apt_dist, dtype: float64

### looking at the distance frequency, it will not be efficient to check every value for correct distance. Therefore I have corrected the only negative value

###  For visibility values The largest reportable metric value is 9999. This value represents a visibility greater than 9000 meters (7 SM or more).

In [23]:
df['Visibility Miles'].describe()

count    40117.000000
mean        11.438884
std         94.176679
min          0.000000
25%         10.000000
50%         10.000000
75%         10.000000
max       9999.000000
Name: Visibility Miles, dtype: float64

In [24]:
#use constraint in counts
df['Visibility Miles'].value_counts().loc[lambda x : x > 500]

10.0    34916
7.0       727
9.0       598
8.0       514
Name: Visibility Miles, dtype: int64

In [25]:
df.loc[df['Visibility Miles'] == 9999]

ev_id  Aircraft_Key Aircraft_Mfg_Mdl_srs acft_category  \
32419  20170819X44220             1         LEARJET-25-D           AIR   
43486  20080502X00587             1          Cessna-500-           AIR   

       total_seats  num_eng date_last_insp  \
32419          NaN      2.0            NaT   
43486          NaN      NaN            NaT   

       Airframe_Hrs_since_last_inspection  Airframe_hrs_total  \
32419                                 NaN                 NaN   
43486                                 NaN                 NaN   

                     oper_name USA_State   oper_country oper_pax_cargo  \
32419              EL MARSHALL       NaN  United States            NaN   
43486  Relton Mews Air Limited       NaN            NaN            PAX   

      type_fly ev_type    ev_date Day Time of Event  ev_year  ev_month  \
32419      NaN     ACC 2017-08-19  Sa      04:15:00     2017         8   
43486     PERS     ACC 2008-03-30  Su      13:38:00     2008         3   

      mid_air latitude longitude Airport Nearest_Airport On or off Airport  \
32419     NaN  103850N  0664431W     NaN             NaN              OFAP   
43486     NaN  512153N  0000040E     NaN             NaN               NaN   

       apt_dist light_cond Sky_condition  Runway_Visibility_Feet  \
32419       0.0       NITE          SCAT                     NaN   
43486       0.0        NaN          SCAT                     NaN   

       Visibility Miles  gust_kts  Total Fatalities  Total Injuries  \
32419            9999.0       0.0                 5             5.0   
43486            9999.0       0.0                 5             5.0   

      Investigating_Agency Weather_Condition  dec_latitude  dec_longitude  \
32419                    O               VMC     10.647222     -66.741943   
43486                    O               VMC     51.364723       0.066666   

                                         Events_Sequence engines Type  \
32419  Enroute-climb to cruise Loss of control in flight           TJ   
43486                                                NaN          NaN   

      engine_Manufacturer  
32419    General Electric  
43486                 NaN

In [26]:
df_Vis = df[df['Visibility Miles']> 00.0]

In [27]:
df_Vis.shape

(40019, 41)

In [28]:
df_Vis

ev_id  Aircraft_Key              Aircraft_Mfg_Mdl_srs  \
0      20080107X00026             1                  PIPER-PA 28-180-   
1      20080107X00026             1                  PIPER-PA 28-180-   
2      20080107X00026             2          Barnard/Stancil-Glastar-   
3      20080107X00026             2          Barnard/Stancil-Glastar-   
4      20080109X00036             1  Micco Aircraft Company-MAC-145B-   
...               ...           ...                               ...   
43881  20101119X80223             1                   Schweizer-269C-   
43895  20101014X52451             1                   ROBINSON-R44-II   
43896  20101014X52451             1                   ROBINSON-R44-II   
43982  20201009102120             1               RANS-S-7 -NO SERIES   
43983  20210706103423             1           QUAD CITY-CHALLENGER-II   

      acft_category  total_seats  num_eng date_last_insp  \
0               AIR          4.0      1.0     2007-04-01   
1               AIR          4.0      1.0     2007-04-01   
2               AIR          2.0      1.0     2007-01-01   
3               AIR          2.0      1.0     2007-01-01   
4               AIR          2.0      1.0     2007-09-01   
...             ...          ...      ...            ...   
43881           AIR          NaN      NaN            NaT   
43895           AIR          NaN      1.0            NaT   
43896           AIR          NaN      1.0            NaT   
43982           AIR          2.0      1.0            NaT   
43983           AIR          NaN      1.0            NaT   

       Airframe_Hrs_since_last_inspection  Airframe_hrs_total  \
0                                    75.0              2865.0   
1                                    75.0              2865.0   
2                                     NaN               340.0   
3                                     NaN               340.0   
4                                     NaN               516.0   
...                                   ...                 ...   
43881                                 NaN                 NaN   
43895                                 NaN                 NaN   
43896                                 NaN                 NaN   
43982                                 NaN                 NaN   
43983                                 NaN                 NaN   

                    oper_name                       USA_State   oper_country  \
0             Timothy Bennett  CALIFORNIA                      United States   
1             Timothy Bennett  CALIFORNIA                      United States   
2               Kenneth Couey  CALIFORNIA                      United States   
3               Kenneth Couey  CALIFORNIA                      United States   
4                         NaN                             NaN            NaN   
...                       ...                             ...            ...   
43881                     NaN                             NaN            NaN   
43895  Elen de Souza Santiago                             NaN            NaN   
43896  Elen de Souza Santiago                             NaN            NaN   
43982                     NaN  IDAHO                           United States   
43983                     NaN  MISSOURI                        United States   

      oper_pax_cargo type_fly ev_type    ev_date Day Time of Event  ev_year  \
0                NaN     PERS     ACC 2008-01-01  Tu      22:30:00     2008   
1                NaN     PERS     ACC 2008-01-01  Tu      22:30:00     2008   
2                NaN     PERS     ACC 2008-01-01  Tu      22:30:00     2008   
3                NaN     PERS     ACC 2008-01-01  Tu      22:30:00     2008   
4                NaN     PERS     ACC 2008-01-01  Tu      22:00:00     2008   
...              ...      ...     ...        ...  ..           ...      ...   
43881            NaN     PERS     ACC 2010-07-25  Su      08:00:00     2010   
43895            NaN     UNK      ACC 2010-0

In [31]:
# use pandas between() function
df_low = df[df['Visibility Miles'].between(50, 9999)]

In [32]:
df_low.shape

(197, 41)

### Going by some websites, max visibility that can be reported is 10 miles but on some manual stations it could be reported up to 70 miles or sometimes higher. In some parts of the world,the visibility is reported in meters,9999 being equivalent to 7 miles. We have 197 values that are above 50 and it can be reasonable assumed that they depict good visibility conditions. No action taken.

### gust_kts ,Max value 999 seems unrealistic. highest recorded value is 220 kns https://en.wikipedia.org/wiki/Wind_speed
### Therefore, all values above it seemed to be erroneous and have been replaced by NaN.


In [33]:
df['gust_kts'].values[df['gust_kts'].values > 220 ] = np.nan

In [37]:
df['gust_kts'].describe()

count    43944.000000
mean         2.957742
std          7.508319
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max         57.000000
Name: gust_kts, dtype: float64

### Runway_Visibility_Feet 
### one negative value observed, cannot be negative, Imputed with NaN


In [35]:
df['Runway_Visibility_Feet'].values[df['Runway_Visibility_Feet'].values < 0 ] = np.nan

In [36]:
df['Runway_Visibility_Feet'].describe()

count      59.000000
mean     2701.474576
std      1787.817590
min         5.000000
25%      1300.000000
50%      2000.000000
75%      4000.000000
max      6000.000000
Name: Runway_Visibility_Feet, dtype: float64

## Country and city columns were inadvertantly ommited from the database. The clean dataframe was exported as xlsx. these coulmns are added using index-match and clean dataset is re-imported

In [38]:
#reimport clean
df= pd.read_excel(r'D:\PythonPrjkt\NTSB_Aviation_accident_Analysis\02_Data\Prepared_Data\NTSB_Aviation_Clean.xlsx',index_col = False)

In [40]:
dt = df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43989 entries, 0 to 43988
Data columns (total 44 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Unnamed: 0                          43989 non-null  int64         
 1   ev_id                               43989 non-null  object        
 2   Aircraft_Key                        43989 non-null  int64         
 3   Aircraft_Mfg_Mdl_srs                43989 non-null  object        
 4   acft_category                       43989 non-null  object        
 5   total_seats                         39237 non-null  float64       
 6   num_eng                             41437 non-null  float64       
 7   date_last_insp                      34193 non-null  datetime64[ns]
 8   Airframe_Hrs_since_last_inspection  13671 non-null  float64       
 9   Airframe_hrs_total                  33305 non-null  float64       
 10  oper_name             

In [49]:
#This result was exported to tabular form for inclusion in the word summary

dt = df.isnull().sum()
xlwriter = pd.ExcelWriter('Multiple.xlsx')
dt.to_excel(xlwriter,sheet_name = 'check',index = True)
xlwriter.close()

# End of the excercise